fix: use modernized and standardized OpenTelemetry when tracing

odeke-em opened the pull request:
This change modernizes trace span attributes by using OpenTelemetry's
semantic conventions that are standardized and allow for much
better common ground adoption by broader systems, even more as
Google Cloud Tracing & Monitoring pushes towards OpenTelemetry more.

With this change we've made the replacement of these fields, directly
with imports from `opentelemetry.semconv.trace.SpanAttributes`, as:

* "db.type"       => DB_SYSTEM aka "db.system"
* "db.url"        => DB_CONNECTION_STRING aka "db.connection_string"
* "db.instance"   => DB_NAME aka "db.name"
* "net.host.name" => NET_HOST_NAME aka "net.host.name"

While here, also updated opentelemetry-(api, sdk) dependencies to use
versions "1.25.0", then opentelemetry-(instrumentation) to "0.46b0"

Also added in an option toggled by environment variable

    ENABLE_EXTENDED_TRACING=true

which allows spans to be annotated with the SQL statement keyed by

    "db.statement"

Fixes #1170
Fixes #1171
Fixes #1173

## Exhibit
With the change made, we now have this trace with attributes visualized from Google Cloud Tracing per
<img width="1434" alt="Screenshot 2024-07-26 at 7 12 42 PM" src="https://github.com/user-attachments/assets/f71df423-9a99-46dd-994f-dfb763416b6c">
<img width="699" alt="Screenshot 2024-07-26 at 7 11 10 PM" src="https://github.com/user-attachments/assets/ae99a24c-8f1f-4487-83e5-04075e42f83c">

odeke-em committed:
fix: use modernized and standardized OpenTelemetry when tracing

This change modernizes trace span attributes by using OpenTelemetry's
semantic conventions that are standardized and allow for much
better common ground adoption by broader systems, even more as
Google Cloud Tracing & Monitoring pushes towards OpenTelemetry more.

With this change we've made the replacement of these fields, directly
with imports from `opentelemetry.semconv.trace.SpanAttributes`, as:

* "db.type"       => DB_SYSTEM aka "db.system"
* "db.url"        => DB_CONNECTION_STRING aka "db.connection_string"
* "db.instance"   => DB_NAME aka "db.name"
* "net.host.name" => NET_HOST_NAME aka "net.host.name"

While here, also updated opentelemetry-(api, sdk) dependencies to use
versions "1.25.0", then opentelemetry-(instrumentation) to "0.46b0"

Also added in an option toggled by environment variable

    ENABLE_EXTENDED_TRACING=true

which allows spans to be annotated with the SQL statement keyed by

    "db.statement"

Fixes #1170
Fixes #1171
Fixes #1173

odeke-em committed:
.github/workflows: toggle and test with EXTENDED_TRACING_ENABLED=true

odeke-em committed:
tracing: use standardized db.system=spanner

odeke-em commented:
@harshachinta is it possible for you to hep me authorize workflows to have the tests running on every push? Thank you.

odeke-em commented:
@harshachinta please see the updated and rebased code.

harshachinta commented on .github/workflows/integration-tests-against-emulator.yaml:
```diff
@@ -24,9 +24,17 @@ jobs:
           GOOGLE_CLOUD_PROJECT: emulator-test-project
           GOOGLE_CLOUD_TESTS_CREATE_SPANNER_INSTANCE: true
+          EXTENDED_TRACING_ENABLED: false
+      - name: Run system tests (with extended tracing)
```
Why do we need a seperate job for extended tracing? What are we validating with the help of this job?

odeke-em replied:
It's to ensure that extended tracing works as expected because there is no other mechanism to test for it.

harshachinta replied:
EXTENDED_TRACING_ENABLED will add sql statements in span attributes right?
I don't think we need to run a job explicitly for that if we don't have any tests specifically to validate this behavior in system tests.

harshachinta commented on docs/opentelemetry-tracing.rst:
```diff
@@ -9,6 +9,7 @@ To take advantage of these traces, we first need to install OpenTelemetry:
 .. code-block:: sh
 
     pip install opentelemetry-api opentelemetry-sdk opentelemetry-instrumentation
+    pip install opentelemetry-exporter-google-cloud
 
     # [Optional] Installs the cloud monitoring exporter, however you can use any exporter of your choice
     pip install opentelemetry-exporter-google-cloud
```
This package is deprecated and will not receive any more updates. Also, this line is already available below and is a duplicate here.
https://pypi.org/project/opentelemetry-exporter-google-cloud/

Replace this with `pip install opentelemetry-exporter-gcp-trace`

harshachinta commented on examples/trace.py:
```diff
@@ -0,0 +1,61 @@
+    sampler = ALWAYS_ON
+    tracerProvider = TracerProvider(sampler=sampler)
+    tracerProvider.add_span_processor(
+        BatchExportSpanProcessor(CloudTraceSpanExporter()))
```
Pass `project_id` to `CloudTraceSpanExporter()`?

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,19 +15,39 @@
     HAS_OPENTELEMETRY_INSTALLED = False
 
 
+EXTENDED_TRACING_ENABLED = os.environ.get('SPANNER_ENABLE_EXTENDED_TRACING', '') == 'true'
```
In addition to env variable, customer should be given an option to set it via an option during client initialization.

odeke-em replied:
Roger that, I am addressing that shortly!

odeke-em replied:
Done!

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,19 +15,39 @@
+EXTENDED_TRACING_ENABLED = os.environ.get('SPANNER_ENABLE_EXTENDED_TRACING', '') == 'true'
+
+
+def annotate_with_sql_statement(span, sql):
```
Where is this getting called?

odeke-em replied:
I removed it, thank you!

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,19 +15,39 @@
+
 @contextmanager
 def trace_call(name, session, extra_attributes=None):
     if not HAS_OPENTELEMETRY_INSTALLED or not session:
```
1. We are using global tracer provider here by default. But we should also give an option for customer to pass customer tracer provider for spanner traces and fallback to global if not provided.
2. Set the instrumenting_module_name to `cloud.google.com/language`, instrumenting_library_version when calling get_tracer()

odeke-em replied:
All done! Thank you @harshachinta!

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -39,15 +59,18 @@ def trace_call(name, session, extra_attributes=None):
-        "db.url": SpannerClient.DEFAULT_ENDPOINT,
-        "db.instance": session._database.name,
-        "net.host.name": SpannerClient.DEFAULT_ENDPOINT,
+        DB_SYSTEM: "spanner",
+        DB_CONNECTION_STRING: SpannerClient.DEFAULT_ENDPOINT,
+        DB_NAME: session._database.name,
+        NET_HOST_NAME: SpannerClient.DEFAULT_ENDPOINT,
```
This can be a breaking change to customers already using these attributes. Lets discuss about this tomorrow.

odeke-em committed:
all: allow injecting a custom tracerProvider else fallback to the global one

odeke-em committed:
docs: use the non-deprecated GCP tracing package

odeke-em committed:
all: fix docs, unify & simplify retrieval of Spanner tracer

Exposes "get_tracer" a function that'll hide the cumbersome
implementation details of the package name and version, retrieving
the tracer uniformly.

odeke-em committed:
allow enable_extended_tracing to be passed in when creating a client

odeke-em committed:
observability: use full namespace cloud.google.com/python/spanner+SPAN_NAME

odeke-em committed:
docs, examples: show how to enable gRPC instrumentation

odeke-em committed:
tracing: properly record the spanner endpoint and show how to enable gRPC

harshachinta commented on docs/opentelemetry-tracing.rst:
```diff
@@ -9,32 +9,84 @@ To take advantage of these traces, we first need to install OpenTelemetry:
+
+.. code:: python
+
+   tracer = tracerProvider.get_tracer('cloud.google.com/python/spanner', SPANNER_LIB_VERSION)
```
As per opentelemetry, this module name `cloud.google.com/python/spanner ` and version `SPANNER_LIB_VERSION ` are used to identify which library is the trace getting generated from. For ex, traces generated in python spanner client will be identified through these variables.

For customers if they want to add custom trace in their application, they will need to use their package name and application version that will help them understand that the custom trace is getting emited from their application.
There is no usecase for customer to retrieve the tracer with these python spanner package names. So you can remove this example here

harshachinta commented on docs/opentelemetry-tracing.rst:
```diff
@@ -9,32 +9,84 @@ To take advantage of these traces, we first need to install OpenTelemetry:
+    )
+
+
+Alternatively you can pass in a tracer provider into the Cloud Spanner initialization
```
Can we mention here that if the tracer provider is. not provided then the python spanner client will use the configured global tracer provider?

harshachinta commented on docs/opentelemetry-tracing.rst:
```diff
@@ -9,32 +9,84 @@ To take advantage of these traces, we first need to install OpenTelemetry:
+        BatchSpanProcessor(CloudTraceSpanExporter())
     )
 
+    o11y = dict(tracer_provider=tracerProvider)
```
can we rename the variable to some meaningful name?

harshachinta commented on examples/grpc_instrumentation_enabled.py:
```diff
@@ -0,0 +1,71 @@
+        BatchSpanProcessor(traceExporter))
+    trace.set_tracer_provider(tracerProvider)
+    # Retrieve the set shared tracer.
+    tracer = tracerProvider.get_tracer('cloud.google.com/python/spanner', spanner.__version__)
```
I dont think this is needed. In this case `QueryInformationSchema` span will be shown as being generated from python spanner client, but it is generated from customer application.

Can you update the module name and version name to application name and version and see how the output changes?
Can you share me the results of both, with change and without change?

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,41 +15,89 @@
+    if not extended_tracing:
+        attributes.pop(DB_STATEMENT, None)
+
+    span_name = SPAN_NAMESPACE + '/' + name
```
What is the trace span output with this change?
I assume the span name will change from `CloudSpanner.CreateSession` to `cloud.google.com/python/spanner/CloudSpanner.CreateSession`. If yes this change is not needed, span name can be as it is and need not be prefixed with namespace.

harshachinta commented on google/cloud/spanner_v1/batch.py:
```diff
@@ -205,7 +205,8 @@ def commit(
             request_options=request_options,
         )
-        with trace_call("CloudSpanner.Commit", self._session, trace_attributes):
+        with trace_call("Batch.commit", self._session, trace_attributes,
```
Can we follow the span naming as per this document?
https://docs.google.com/document/d/1C-2WbnKfAtl-GgOebVNYv0RdYFkb_UfyzagwJOQgxAs/edit?tab=t.0#bookmark=id.ctslkgwt101t

I would suggest to keep the span naming changes in a seperate PR. This will make PR small and easy to review. We will focus on getting this PR out.

harshachinta replied:
@odeke-em 
I am in favour of not changing the span names since that could introduce breaking changes.

harshachinta replied:
I would also suggest if you can take a look at the python spanner code and suggest any additional spans that could help customer for better debugging. 
You can also look at this document to check the existing spans in other languages to get some idea.
https://docs.google.com/document/d/1C-2WbnKfAtl-GgOebVNYv0RdYFkb_UfyzagwJOQgxAs/edit?tab=t.0#bookmark=id.ctslkgwt101t

harshachinta commented on google/cloud/spanner_v1/session.py:
```diff
@@ -63,12 +63,13 @@ class Session(object):
     _transaction = None
 
-    def __init__(self, database, labels=None, database_role=None):
+    def __init__(self, database, labels=None, database_role=None, observability_options=None):
```
Just to clarify is this method used internally or by customers?
If only used internally, I don't see that we are passing observability_options to this method that is received through spanner client.

harshachinta commented on google/cloud/spanner_v1/__init__.py:
```diff
@@ -88,6 +89,7 @@
     # google.cloud.spanner_v1
     "__version__",
     "param_types",
+    "get_tracer",
```
Why do we need this here?

harshachinta replied:
This is not something that customer will need. We can remove this

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,41 +15,89 @@
+    it'll fall back to the global tracer provider and use this library's specific semantics.
+    """
+    if tracer_provider:
+        return tracerProvider.get_tracer(TRACER_NAME, TRACER_VERSION)
```
```suggestion
        return tracer_provider.get_tracer(TRACER_NAME, TRACER_VERSION)
```

harshachinta replied:
I tried to run one sample and see that `otel.scope.name` is not getting populated.
![image](https://github.com/user-attachments/assets/a9a7cb93-6046-4b3f-9751-9d72df3163a5)

It should look something like below (this was in java)
![image](https://github.com/user-attachments/assets/36cfc5bc-a937-40c4-9dc9-46ef1ba9e2a3)

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,41 +15,89 @@
-    tracer = trace.get_tracer(__name__)
+    tracer = None
+    if observability_options:
+        tracerProvider = observability_options.get('tracer_provider', None)
```
variable should be lower case

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,41 +15,89 @@
-        "db.instance": session._database.name,
-        "net.host.name": SpannerClient.DEFAULT_ENDPOINT,
+        DB_SYSTEM: "spanner",
+        DB_CONNECTION_STRING: spanner_endpoint
```
```suggestion
        DB_CONNECTION_STRING: spanner_endpoint,
```

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -15,41 +15,89 @@
+    from opentelemetry.semconv.trace import SpanAttributes
 
     HAS_OPENTELEMETRY_INSTALLED = True
+    DB_SYSTEM = SpanAttributes.DB_SYSTEM
```
1. We are changing attribute conventions here. As discussed earlier do we need to make this change under a seperate option to avoid breaking changes?
2. I also see that we might need to change span names to be consistent with this [document](https://docs.google.com/document/d/1C-2WbnKfAtl-GgOebVNYv0RdYFkb_UfyzagwJOQgxAs/edit?tab=t.0#bookmark=id.ctslkgwt101t), which is anyhow a breaking change.

So should we make these changes directly or instead protect them under an option?
My opinion is since we are refactoring observability portion, lets make all changes directly and warn the customers in release notes. Having a seperate option will make code messier.
@surbhigarg92 Can you give your thoughts on this?

harshachinta replied:
@odeke-em 
1. For attribute conventions lets enable it under a seperate flag to avoid breaking changes.
2. Lets not change any existing span names as they could bring in a breaking change. Instead can you help add additional spans that could help in debugging.

odeke-em committed:
tracing: add otel.scope.name,version for every span

odeke-em committed:
plumb in observability_options into _SessionWrapper constructor

odeke-em committed:
no need to expose get_tracer

odeke-em committed:
revert span name changes

odeke-em committed:
update docs/opentelemetry-tracing

odeke-em committed:
tracing: plumb in more otel.scope.* + fix retrieval from global tracer provider

odeke-em commented:
Superseded by https://github.com/googleapis/python-spanner/pull/1199
